MetaMask/metamask-extension
# Fix nonces in smart swaps requiring approval

## **Description**

When performing a smart swap that requires approval, both transactions have the same nonce since the core `TransactionController` did not support external pending transactions in the nonce calculation.

This support has been added to the core `TransactionController` through the `getExternalPendingTransactions` callback.

See [core PR ](https://github.com/MetaMask/core/pull/3587) for reference.

## **Related issues**

Fixes #22093 

## **Manual testing steps**

1. Ensure `Smart swap` is enabled in `Swaps > Settings`.
2. Perform a swap from a token that has not been approved.
3. Ensure both smart transactions are successful and have sequential nonces.

## **Screenshots/Recordings**



### **Before**



### **After**



## **Pre-merge author checklist**


## **Pre-merge reviewer checklist**
